src/algebra/hierarchy_design.lean

```diff
@@ -0,0 +1,154 @@
+/-
+Copyright (c) 2021 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison, Eric Weiser
+-/
+import tactic.doc_commands
+
+/-!
+# Documentation of the algebraic hierarchy
+
+-/
+
+/--
+# The algebraic hierarchy
+
+In any theorem proving environment,
+there are difficult decisions surrounding the design of the "algebraic hierarchy".
+
+There is a danger of exponential explosion in the number of gadgets,
+especially once interactions between algebraic and order/topological/etc structures are considered.
+
+In mathlib, we try to avoid this by only introducing new algebraic typeclasses either
+1. when there is "real mathematics" to be done with them, or
+2. when there is a meaninful gain in simplicity by factoring out a common substructure.
+
+(As examples, at this point we don't have `loop`, or `unital_magma`,
+but we do have `lie_submodule` and `topological_field`!)
```

Comment:
Would `group_with_zero` be a good example to illustrate 2) here?

Reply:
Yes! Want to add a paragraph?